### ISSUE TITLE
model中的mapping

### ISSUE THREAD

**Issue by zuobaifu** [wjh892521292/Ord2Seq#2]
您好，打扰啦，pvttrans和vggtrans模型中为什么对于相同的类别数所对应的mapping不一样，mapping是如何计算的，和论文里好像不太一样

**Comment by wjh892521292** [comment]
论文里前面数据集的mapping应该是参考vggtrans的，dr的会不一样。这里的mapping是根据dichotomy tree的划分来进行标序的，同样的数值表明是一个集合，也就是同样的父节点。mapping可以根据不同的数据集自己设置，相当于定义dichotomy tree的划分方式。

**Comment by zuobaifu** [comment]
很高兴您的回复，想询问一下您这里的mapping操作是不是对应论文里Label Embedding中的h函数，如果是的话那么对于相同的类别数，pvttrans中的mapping与vggtrans中的mapping对于类别数为5的时候为什么不一样？是因为在pvttrans和vggtrans中dichotomy tree的划分不一样嘛，如果使用的话是按照论文给的示例图的方式划分还是怎么划分？期待您的回复。

**Comment by wjh892521292** [comment]
抱歉忘记回复邮件了。 mapping操作是对应论文里Label Embedding中的h函数。 pvttrans中的mapping与vggtrans中的mapping对于类别数为5的时候不一样是因为面向的是不同的数据集（类别都为5但是数据量和分布不同），因此设定了不同的dichotomy tree的划分方式，尽可能让树的左右儿子节点样本量平衡（正负样本平衡）。 总而言之，划分方式可以根据数据集的分布自定义。


892521292
***@***.***



&nbsp;




------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "wjh892521292/Ord2Seq"                                                                                    ***@***.***&gt;;
发送时间:&nbsp;2023年11月8日(星期三) 晚上7:23
***@***.***&gt;;
***@***.******@***.***&gt;;
主题:&nbsp;Re: [wjh892521292/Ord2Seq] model中的mapping (Issue #2)





 
很高兴您的回复，想询问一下您这里的mapping操作是不是对应论文里Label Embedding中的h函数，如果是的话那么对于相同的类别数，pvttrans中的mapping与vggtrans中的mapping对于类别数为5的时候为什么不一样？是因为在pvttrans和vggtrans中dichotomy tree的划分不一样嘛，如果使用的话是按照论文给的示例图的方式划分还是怎么划分？期待您的回复。

**Comment by zuobaifu** [comment]
您好，在论文里Masked Decision部分，第一：在计算y_prob时是mask*sigmoid(y_out)，代码里直接只用的是带有sigmoid的BCE损失函数那么最后计算的y_prob是sigmoid(mask*y_out)，与论文描述就不一致啦；第二：Masked Decision部分使用到的y_mht是根据y_prob得到的嘛？最近在复现这篇论文，指标一直都没有跑到论文中所展示的指标很是困惑。

**Comment by wjh892521292** [comment]
1.mask只在train阶段需要使用，所以if self.training里 out先乘以mask， 再和label（y_mht）计算损失。这部分和论文描述是一致的。
2. y_mht是标签，根据设置的mapping可以得到相应的y_mht，也就是根据dictomy tree为每一层确定正负样本(1和0）。&nbsp;








892521292
***@***.***



&nbsp;




------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "wjh892521292/Ord2Seq"                                                                                    ***@***.***&gt;;
发送时间:&nbsp;2023年11月22日(星期三) 晚上6:17
***@***.***&gt;;
***@***.******@***.***&gt;;
主题:&nbsp;Re: [wjh892521292/Ord2Seq] model中的mapping (Issue #2)





 
您好，在论文里Masked Decision部分，第一：在计算y_prob时是masksigmoid(y_out)，代码里直接只用的是带有sigmoid的BCE损失函数那么最后计算的y_prob是sigmoid(masky_out)，与论文描述就不一致啦；第二：Masked Decision部分使用到的y_mht是根据y_prob得到的嘛？最近在复现这篇论文，指标一直都没有跑到论文中所展示的指标很是困惑。

**Comment by zuobaifu** [comment]
![image](https://github.com/wjh892521292/Ord2Seq/assets/57834082/6773d610-a606-4914-916f-e2b0c7fbcc44)
您好，这一部分描述的只是针对于train阶段嘛？对于测试阶段，取每一步的softmax(y_out)的最大值，作为预测，然后作为下一步的输入，这里不涉及mask，不知道是否可以这样理解？
![image](https://github.com/wjh892521292/Ord2Seq/assets/57834082/68d02311-cb80-4fa2-8831-e25e59802a24)
论文描述的是先进行sigmoid函数对于y_out,然后再与mask相乘，再与label计算损失；但是代码里写的是y_out先乘以mask，然后再和label计算损失，因为使用的是带有sigmoid函数的BCE，所以就相当于是对y_out与mask的乘积的结果进行sigmoid函数然后再与label计算损失，不知道我这样理解是否正确。
